REST API: Web Authorization

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8599 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR implements the Application Password's web authorization flow.

I suggest reviewing the PR by commits, it'll help understanding the reasoning for each decision.

Previously, REST API login only supports native login, as outlined in the diagram below (click for bigger version):

![REST API login flow i1 complete](https://user-images.githubusercontent.com/266376/227841781-1378e9ce-fd25-486e-852f-6253cfdad956.png)


Behind the scene, this process requires authenticating via `wp-login.php` to get the nonce required to generate the application password. With this process, we're seeing a decent amount of login failures caused by inaccessible `wp-login.php` (due to security plugins or server settings, or other causes).

This PR is meant to help with those login failures by adding an alternative login method using webview, if a merchant gets login failure using the native login.

--- 
#### This PR adds the following:
1. A dialog message once the native REST API login fails,
2. An option in the dialog to login via wp-admin, using the authentication URL provided by WP's REST API,
3. A webview to allow such login action,
4. Once webview authentication is completed, handle the automatic redirect back into the app and log the user in.
5. Error handlings.

This expands the diagram into something like below, with webview login added to the flow starting from the login error on top right:

![REST API login flow complete simplified](https://user-images.githubusercontent.com/266376/227835153-d0139884-4845-4b9c-8c91-e1baf6ce3114.png)


---

### Not in this PR:
1. Unit tests related to this feature. To keep the PR small-ish, these are added separately in https://github.com/woocommerce/woocommerce-android/pull/8651
2. Tracks. To keep the PR small-ish, these are added separately.
3. Handling for this edge case issue #8644. It will be tackled later on.

--- 

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

#### Success path: ####

1. Have a test .org WooCommerce site with no Jetpack,
2. Replicate the native REST API login error by doing the following: 
    1. Install **"WP Hide & Security Enhancer"** plugin and activate it,
    2. Go to **"WP Hide"** -> **"Hide > Admin"** in the menu,
    3. On the **"New wp-login.php"** section, enter a custom login url (e.g: "entering.php"),
    4. On the **"Block default wp-login.php"** section, select "Show", then select "Yes"
    5. Click "Save"
3. Start the app,
4. Enter the test site's URL, then enter credentials, and wait until the login failure dialog happens,
5. Select "Try again with wp-admin page",
6. Wait until webview loads, and proceed with the login process,
7. When wp-admin asks you to approve the application passwords connection, please approve it,
8. Wait and confirm that the app logs you in properly.

#### Failure path 1 : Application passwords authentication rejected ####
1. Follow success path steps 1-6,
2. When wp-admin asks you to approve the application passwords connection, please choose "No, I do not approve of this connection",
3. Confirm that the Log in screen is now loaded with a Snackbar explaining the error.


#### Failure path 2: Application passwords disabled ####
This for the case where both `wp-login.php` is inaccessible, and application password is disabled.
1. Follow success path steps 1-2, 
3. In wp-admin, install and activate "Disable Application Passwords" plugin,
4. Follow success path 3-5,
6. Confirm that an error page is shown saying that application password is not available.


### Video

Check video below for a successful login example:

https://user-images.githubusercontent.com/266376/227840023-e303794d-3381-412d-810d-53313a427762.mov

---

Check the screenshots below for the failure cases:

| Application passwords not approved  | Application passwords disabled |
|-|-|
|  ![application password not approved](https://user-images.githubusercontent.com/266376/227840897-4a4cd4ad-4177-4765-b0f2-ebf5f0425e16.png) | ![Screenshot_20230327_112531](https://user-images.githubusercontent.com/266376/227840903-14568ded-c2ec-4cfb-a088-041753c00bfb.png) |


- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Remove the non-used parcelization

hichamboushaba committed:
Introduce a state property

hichamboushaba committed:
Move public functions to the top

hichamboushaba committed:
Add logic to map ViewState from states

This includes fetching the Site if needed when the web authorization state is enabled

hichamboushaba committed:
Update current native login screen to the new changes

hichamboushaba committed:
Add a button to the error dialog to start the web authorization

hichamboushaba committed:
Load the webview when starting the web authorization

hichamboushaba committed:
Handle the post-authorization flow

hichamboushaba committed:
Reset fetchedSiteId when dismissing the Web Authorization flow

hichamboushaba committed:
Handle retrying after the application passwords error screen

hichamboushaba committed:
Unify the site fetching logic between both flows

hichamboushaba committed:
Add a todo comment for validating the placement of the web login button

hichamboushaba committed:
Use custom dialog messages for different flows

hafizrahman committed:
Bump fluxC.

Especially to take into account the fix in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2686

I'm pointing to a commit several times later than that particular fix, though.

hafizrahman committed:
Add: Compose previews.

hafizrahman committed:
Add: Error handling when user refuses to approve application password addition in wp-admin.

hafizrahman committed:
Synchronize buttons text and placement with iOS version.

https://github.com/woocommerce/woocommerce-ios/pull/9187

We do add "Need more help?" while iOS doesn't, but otherwise it's the same.

ParaskP7 committed:
Merge pull request #8647 from woocommerce/release/12.9

Merge release/12.9-rc-1 into trunk (`beta`)

hafizrahman committed:
Fix: LoginSiteCredentialsViewModel unit tests.

hafizrahman committed:
Merge branch 'trunk' into issue/8599-application-passwords-web-auhorization

hafizrahman committed:
Add: Unit test for case displaying native login on init.

hafizrahman committed:
Add: Unit test for case displaying webview login.

hafizrahman committed:
Detekt fixes.

hafizrahman committed:
Add: Unit test for case successful webview login.

hafizrahman committed:
Add: Unit test for case rejected webview login.

hafizrahman committed:
Simplify variable naming.

hafizrahman committed:
Remove unneeded mocking.

hafizrahman committed:
Improve mocking.

hafizrahman committed:
Merge branch 'issue/8599-application-passwords-web-auhorization' into issue/8599-unit-tests-for-application-passwords-web-authorization

hafizrahman committed:
Add: Unit test for case app pwd disabled and wp-login.php inaccessible.

hafizrahman committed:
Detekt fixes.

hafizrahman commented:
Setting as "Do not merge" for now until we decide whether to put this into beta or the next release.

hafizrahman committed:
Add: tracking keys related to REST API via webview- logins.

hafizrahman committed:
Add: tracking for webview shown.

hafizrahman committed:
Add: tracking for application password connection rejected.

hafizrahman committed:
Add: tracking for application password connection accepted.

hafizrahman committed:
Add: tracking for application password auth URL not available.

hafizrahman committed:
Move tracking for application password auth URL not available to a better spot.

hafizrahman committed:
Remove unneeded event.

This is already tracked in a different way, see https://github.com/woocommerce/woocommerce-android/pull/8618/files?diff=split&w=0#diff-6041eda7bbcd79b8879ec4c62d7dcda35fcd19e039a58ae32a9792fb20cca5aeR292-R296

0nko committed:
Merge pull request #8651 from woocommerce/issue/8599-unit-tests-for-application-passwords-web-authorization

[REST API] Web Authorization feature unit tests

hafizrahman committed:
Merge branch 'issue/8599-application-passwords-web-auhorization' into issue/8599-tracks-for-application-passwords-web-authorization

hafizrahman committed:
Improve the way we track webview shown.

Previously, since it's inside a combine, the tracking can be done multiple times just for a single webview shown event.

hafizrahman committed:
Add: release note for webview-based REST API login.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/sitecredentials/LoginSiteCredentialsViewModel.kt:
```diff
@@ -66,22 +77,18 @@ class LoginSiteCredentialsViewModel @Inject constructor(
-            errorDialogMessage = errorDialog
-        )
+    @OptIn(ExperimentalCoroutinesApi::class)
+    val viewState = state.flatMapLatest {
```
Np, but the field `state` I'd maybe rename it to `initialState` or something similar to clearly be able to differentiate from the actual `viewState`. 

Very nice usage of Flows API in this view model.

JorgeMucientes commented:
Outstanding job here @hichamboushaba and @hafizrahman 🥇. This wasn't an easy task to deal with! The code looks great and works like a charm for the different testing scenarios. 

I just found a minor corner case where the UI state for `LoginSiteCredentialsViewModel` is not restored properly. Basically, when clicking on arrow back from "Application passwords disabled" error screen. See the screen recording: 


https://user-images.githubusercontent.com/2663464/228500586-be90e003-11ec-4311-9d5b-88364da9d6fe.mp4

Since it's a minor thing, I wouldn't block the PR because of that. So, approved. Let's merge it after the tracking PR is merged into this branch.

hafizrahman commented:
@JorgeMucientes Thanks for the review and bug report! Great minds think alike, I've also seen that bug and written about it in #8644. 

I'll fix it in the next update since I think it's quite an edge case. It looks awkward but hopefully not many users disable both wp-login.php access and application passwords at the same time. In the meanwhile they also can still tap the "X" button to close it. Again, very awkward, but not end of the world bug.

hafizrahman committed:
Fix: Move tracking for "webview shown" to be triggered by "Try again with wp-admin" click.

This ensures that it is only ever called once per view.

hafizrahman committed:
Fix: Remove onWebViewShown() usage for tracking, as it's no longer needed.

hafizrahman committed:
Add: Option onPageFinished parameter to WCWebView.

This might seem close in usage with onLoadResource(), but the main purpose here is to have something that calls the function parameter only once, unlike `onUrlLoaded` that gets called twice by `onLoadResource` and `doUpdateVisitedHistory`.

See https://github.com/woocommerce/woocommerce-android/pull/8660/files#r1151742457 for more context.

hafizrahman committed:
Add: Make use of `onPageFinished` to ensure `onWebAuthorizationUrlLoaded` is not called twice each time.

This ensures that tracking inside it only gets called once.

JorgeMucientes commented:
> I'll fix it in the next update since I think it's quite an edge case.

Agreed! Sounds good. 

Btw while re-testing the feature I found a similar corner case.  For "Failure path 2: Application passwords disabled" from the PR description: 
- Try to log in with app paswords disabled
- See error page is shown saying that the application password is not available.
- Go to `wc-admin` disable the plugin "Disable Application Passwords"
- Click "Retry" button in the app from the application passwords disabled error screen 
- Confirm the blank screen


https://user-images.githubusercontent.com/2663464/228665131-0015a032-acfe-48b3-9e9e-cda9c79cd851.mp4

This is also a non blocking corner case but maybe you want to keep track of it.

hafizrahman committed:
Merge pull request #8660 from woocommerce/issue/8599-tracks-for-application-passwords-web-authorization

[REST API] Analytics for application passwords web authorization

hafizrahman commented:
Just to note that between these last few commits, I've seen check failures related to e2e test, especially in order creation flow. I did some testing and couldn't replicate it. There is nothing in this PR that relates to that issue, and in last commit https://github.com/woocommerce/woocommerce-android/pull/8618/commits/7a71d9ce01979230dd1949352d9c87e69c0b3ee8 the failure disappeared on its own without any specific fixing from this PR. So in case this happens again in the next commits, I'd say it should be OK to ignore.

hafizrahman committed:
Fix: Make webview login retrying work if initially app pwd is disabled, and then enabled before retrying.

This is to fix this bug report: https://github.com/woocommerce/woocommerce-android/pull/8618#issuecomment-1489306887

The `State.RetryWebAuthorization` state is needed to ensure the `state` emits a new value, thus triggering a new webview reload. We can't re-use `State.WebAuthorization` in this retrying case because the state is already `WebAuthorization` at this point.

hafizrahman commented:
> Btw while re-testing the feature I found a similar corner case. For "Failure path 2: Application passwords disabled" from the PR description:

@JorgeMucientes nice find, thanks for the report! I've added a fix in https://github.com/woocommerce/woocommerce-android/pull/8618/commits/5df0192433cd660aa220855226655da1b35a21c0
